HW9: Saravayskiy Mike

maiksaray opened the pull request:
Implemented the Stats thingy in the following way:

We call private method with the accumulator that tracks already visited (counted) users and add the root user there right away.
Once we see new id in followers list - we add to the accumulator and call the same method recursively.
That allows to visit each user only once and safely check that user was already visited only once
To access network only once per userstats/followers list - we have ConcurrentHashMaps as caches.

There is no waiting, no additional overhead for parallelization, no nothing ugly (except some java syntax)

BUT I STILL COULDN'T BEAT 2 threads vs 1 thread test!
![image](https://user-images.githubusercontent.com/2794213/82367253-2e040a80-9a1c-11ea-9123-1d0dca29086b.png)
`5934963800 > 4660506200` tho

maiksaray committed:
Implemented FollowersStatsImpl, two threads test still fails tho, not sure if I can fix this, code seems to be working ok

maiksaray committed:
Polished the code, optimized computing so that we check id being visited only once as fast as possible.
Still can't beat two threads test.
There are some leftovers from another approach so that if I'm wrong with ConcurrentHashMap.putIfAbsent()==null I still have the solutions

maiksaray committed:
Properly implemented parallel counting, two threads are now really faster, but NOT ENOUGHT FASTER

maiksaray commented:
I have no idea why the build failed, close-opening the PR

maiksaray commented:
Hmm

Walingar commented:
It failed because of `test 2 threads should be faster than 1 thread` test ;)

Walingar commented on src/main/java/impl/network/FollowersStatsImpl.java:
```diff
@@ -0,0 +1,86 @@
+            }
+
+            for (var future : followersFutures) {
+                thisUserValue += future.get();
```
It is not asynchronous way. Here you wait for futures, but you shouldn't

maiksaray committed:
OK, I'm completely broken by all these thenCompose/Combine.
I've spent 4hrs trying to understand how can I return Future value without get() from inside thenCombine()!
LITERALLY FOUR HOURS.

But in the end I think this looks nice =)

Maybe you get used to these once you've spent some time getting into all of this...
